src/mon/NVMeofGwMap.cc

```diff
@@ -0,0 +1,642 @@
+{
+    switch (state) {
+    case gw_states_per_group_t::GW_WAIT_BLOCKLIST_CMPL:
+    {
+        int timer_val = get_timer(gw_id, group_key, grpid);
+        NvmeGwMonState& gw_map = created_gws[group_key][gw_id];
+            if(gw_map.blocklist_data[grpid].osd_epoch <= last_osd_epoch){
+                dout(4) << "is-failover: " << gw_map.blocklist_data[grpid].is_failover << " osd epoch changed from " << gw_map.blocklist_data[grpid].osd_epoch << " to "<< last_osd_epoch
+                                       << " Ana-grp: " << grpid  << " timer:" << timer_val << dendl;
+                gw_state.active_state(grpid);                   // Failover Gw still alive and guaranteed that
+                cancel_timer(gw_id, group_key, grpid);          // ana group wouldnt be taken back  during blocklist wait period
+                map_modified = true;
+            }
+            else{
+                dout(20) << "osd epoch not changed from " <<  gw_map.blocklist_data[grpid].osd_epoch << " to "<< last_osd_epoch
+                                                   << " Ana-grp: " << grpid  << " timer:" << timer_val << dendl;
+            }
+    }
+    break;
+
+    default:
+        break;
+    }
+}
+
+ void NVMeofGwMap::fsm_handle_gw_down(const NvmeGwId &gw_id, const NvmeGroupKey& group_key,   gw_states_per_group_t state, NvmeAnaGrpId grpid,  bool &map_modified)
+ {
+    switch (state)
+    {
+        case gw_states_per_group_t::GW_STANDBY_STATE:
+        case gw_states_per_group_t::GW_IDLE_STATE:
+            // nothing to do
+            break;
+
+        case gw_states_per_group_t::GW_WAIT_BLOCKLIST_CMPL:
+        {
+            cancel_timer(gw_id, group_key, grpid);
+        }break;
+
+        case gw_states_per_group_t::GW_WAIT_FAILBACK_PREPARED:
+            cancel_timer(gw_id, group_key,  grpid);
+
+            for (auto& gw_st: created_gws[group_key]) {
+                auto& st = gw_st.second;
+                if (st.sm_state[grpid] == gw_states_per_group_t::GW_OWNER_WAIT_FAILBACK_PREPARED) { // found GW   that was intended for  Failback for this ana grp
+                    dout(4) << "Warning: Outgoing Failback when GW is down back - to rollback it"  <<" GW "  <<gw_id << "for ANA Group " << grpid << dendl;
+                    st.standby_state(grpid);
+                    map_modified = true;
+                    break;
+                }
+            }
+            break;
+
+        case gw_states_per_group_t::GW_OWNER_WAIT_FAILBACK_PREPARED:
+            // nothing to do - let failback timer expire
+            break;
+
+        case gw_states_per_group_t::GW_ACTIVE_STATE:
+        {
+            find_failover_candidate( gw_id, group_key, grpid, map_modified);
+        }
+        break;
+
+        default:{
+            ceph_assert(false);
```

Comment:
These states are provided directly by the external daemons, right? We can't assert on that unvalidated input.